workflow/scripts/add_district_heating_subnodes.py

```diff
@@ -0,0 +1,421 @@
+                    f"{subnode['cluster']} urban central": f"{subnode['cluster']} {subnode['Stadt']} urban central"
+                },
+                regex=True,
+            )
+            .set_index("Link")
+        )
+        n.madd("Link", links.index, **links)
+
+        # Add heat pumps to subnode
+        heat_pumps = (
+            n.links.filter(
+                regex=f"{subnode['cluster']} urban central.*heat pump", axis=0
+            )
+            .reset_index()
+            .replace(
+                {
+                    f"{subnode['cluster']} urban central": f"{subnode['cluster']} {subnode['Stadt']} urban central"
+                },
+                regex=True,
+            )
+            .set_index("Link")
+        ).drop("efficiency", axis=1)
+        heat_pumps_t = n.links_t.efficiency.filter(
+            regex=f"{subnode['cluster']} urban central.*heat pump"
+        )
+        heat_pumps_t.columns = heat_pumps_t.columns.str.replace(
+            f"{subnode['cluster']} urban central",
+            f"{subnode['cluster']} {subnode['Stadt']} urban central",
+        )
+        n.madd("Link", heat_pumps.index, efficiency=heat_pumps_t, **heat_pumps)
+
+        # Add heat vent to subnode
+        n.madd(
+            "Generator",
+            [f"{name} heat vent"],
+            bus=name,
+            location=f"{subnode['cluster']} {subnode['Stadt']}",
+            carrier="urban central heat vent",
+            p_nom_extendable=True,
+            p_min_pu=-1,
+            p_max_pu=0,
+            unit="MWh_th",
+        )
+
+    return
+
+
+def extend_cops(cops: xr.DataArray, subnodes: gpd.GeoDataFrame) -> xr.DataArray:
+    """
+    Extend COPs (Coefficient of Performance) by subnodes mirroring the timeseries of the corresponding
+    mother node.
+
+    Parameters:
+    cops (xr.DataArray): DataArray containing COP timeseries data.
+    subnodes (gpd.GeoDataFrame): GeoDataFrame containing information about district heating subnodes.
+
+    Returns:
+    xr.DataArray: Extended DataArray with COP timeseries for subnodes.
+    """
+    cops_extended = cops.copy()
+
+    # Iterate over the DataFrame rows
+    for _, subnode in subnodes.iterrows():
+        cluster_name = subnode["cluster"]
+        city_name = subnode["Stadt"]
+
+        # Select the xarray entry where name matches the cluster
+        selected_entry = cops.sel(name=cluster_name)
+
+        # Rename the selected entry
+        renamed_entry = selected_entry.assign_coords(name=f"{cluster_name} {city_name}")
+
+        # Combine the renamed entry with the extended dataset
+        cops_extended = xr.concat([cops_extended, renamed_entry], dim="name")
+
+    # Change dtype of the name dimension to string
+    cops_extended.coords["name"] = cops_extended.coords["name"].astype(str)
+
+    return cops_extended
+
+
+def extend_heating_distribution(
+    existing_heating_distribution: pd.DataFrame, subnodes: gpd.GeoDataFrame
+) -> pd.DataFrame:
+    """
```

Comment:
Could we write the docstring in np. style?

Reply:
Done